cpp/src/arrow/device.h

```diff
@@ -150,9 +150,10 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   /// @param release_fn a function to call during destruction, `nullptr` or
   ///        a no-op function can be passed to indicate ownership is maintained
   ///        externally
-  virtual Result<std::shared_ptr<Stream>> WrapStream(
-      [[maybe_unused]] void* device_stream,
-      [[maybe_unused]] Stream::release_fn_t release_fn) {
+  virtual Result<std::shared_ptr<Stream>> WrapStream(void* device_stream,
+                                                     Stream::release_fn_t release_fn) {
+    ARROW_UNUSED(device_stream);
+    ARROW_UNUSED(release_fn);
```

Comment:
Can we use `ARROW_ARG_UNUSED()` by adding `ARROW_ARG_UNUSED(x)=x` to Doxygen's `PREDEFINED`?

(`ARROW_ARG_UNUSED(x)=x` is added and the list is sorted.)

```diff
diff --git a/cpp/apidoc/Doxyfile b/cpp/apidoc/Doxyfile
index e19c933cd4..5be93032c0 100644
--- a/cpp/apidoc/Doxyfile
+++ b/cpp/apidoc/Doxyfile
@@ -2168,16 +2168,17 @@ INCLUDE_FILE_PATTERNS  =
 
 PREDEFINED             = __attribute__(x)= \
                          __declspec(x)= \
-                         PARQUET_EXPORT= \
-                         GANDIVA_EXPORT= \
-                         ARROW_EXPORT= \
                          ARROW_ACERO_EXPORT= \
+                         ARROW_ARG_UNUSED(x)=x \
+                         ARROW_DEPRECATED(x)= \
                          ARROW_DS_EXPORT= \
                          ARROW_ENGINE_EXPORT= \
+                         ARROW_EXPORT= \
+                         ARROW_EXTERN_TEMPLATE= \
                          ARROW_FLIGHT_EXPORT= \
                          ARROW_FLIGHT_SQL_EXPORT= \
-                         ARROW_EXTERN_TEMPLATE= \
-                         ARROW_DEPRECATED(x)=
+                         GANDIVA_EXPORT= \
+                         PARQUET_EXPORT=
 
 # If the MACRO_EXPANSION and EXPAND_ONLY_PREDEF tags are set to YES then this
 # tag can be used to specify a list of macro names that should be expanded. The
```

See also: https://www.doxygen.nl/manual/preprocessing.html

Reply:
I can get this to work with smaller examples but no such luck yet with the Arrow build. Need to investigate further

Reply:
After some more digging it seems like what you suggested works for Doxygen to generate the appropriate output, but clang does not handle this well. I think this issue upstream is related https://github.com/llvm/llvm-project/issues/77489

Reply:
Here's an MRE.

test.cc:

```cpp
#define ARG_UNUSED(x)

/// @brief A function
///
/// @param foo - this is not really used
void SomeFunction(void ARG_UNUSED(foo)) {
  return;
}
```

Doxyfile:

```
PROJECT_NAME      = foo
INPUT             = test.cc
GENERATE_HTML     = NO
GENERATE_XML      = YES
XML_OUTPUT        = test-xml
MACRO_EXPANSION   = YES
PREDEFINED        = ARG_UNUSED(x)=x
```

`doxygen -d Preprocessor` output:

```
Doxygen version used: 1.9.1
Searching for include files...
Searching for example files...
Searching for images...
Searching for dot files...
Searching for msc files...
Searching for dia files...
Searching for files to exclude
Searching INPUT for files to process...
Reading and parsing tag files
Parsing files
Preprocessing /home/willayd/Desktop/test.cc...
Preprocessor output of /home/willayd/Desktop/test.cc (size: 134 bytes):
---------
00001 #define ARG_UNUSED(x)
00002 
00003 /// @brief A function
00004 ///
00005 /// @param foo - this is not really used
00006 void SomeFunction(void  foo ) {
00007   return;
00008 }
00009 
---------
Macros accessible in this file (/home/willayd/Desktop/test.cc):
---------
ARG_UNUSED 
---------
```

`clang++ test.cc -Wdocumentation -shared` output:

```
test.cc:5:12: warning: parameter 'foo' not found in the function declaration [-Wdocumentation]
/// @param foo - this is not really used
           ^~~
1 warning generated.
```

Reply:
Oh... How about this?

```diff
diff --git a/cpp/cmake_modules/SetupCxxFlags.cmake b/cpp/cmake_modules/SetupCxxFlags.cmake
index d56609c123..ea357b4779 100644
--- a/cpp/cmake_modules/SetupCxxFlags.cmake
+++ b/cpp/cmake_modules/SetupCxxFlags.cmake
@@ -314,6 +314,7 @@ if("${BUILD_WARNING_LEVEL}" STREQUAL "CHECKIN")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wall")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wextra")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wdocumentation")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -DARROW_WARN_DOCUMENTATION")
     if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
       # size_t is 32 bit in Emscripten wasm32 - ignore conversion errors
       set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wno-shorten-64-to-32")
diff --git a/cpp/src/arrow/util/macros.h b/cpp/src/arrow/util/macros.h
index d80828869b..6faf25b3c9 100644
--- a/cpp/src/arrow/util/macros.h
+++ b/cpp/src/arrow/util/macros.h
@@ -67,7 +67,11 @@
 // [5] J. Doerfert et al. 2019. "Performance Exploration Through Optimistic Static
 //     Program Annotations". https://github.com/jdoerfert/PETOSPA/blob/master/ISC19.pdf
 #define ARROW_UNUSED(x) (void)(x)
+#ifdef ARROW_WARN_DOCUMENTATION
+#define ARROW_ARG_UNUSED(x) (x)
+#else
 #define ARROW_ARG_UNUSED(x)
+#endif
 #if defined(__GNUC__)  // GCC and compatible compilers (clang, Intel ICC)
 #define ARROW_NORETURN __attribute__((noreturn))
 #define ARROW_NOINLINE __attribute__((noinline))
```

Reply:
Just pushed up this change